libraries/AP_WayBack/AP_WayBack.cpp

```diff
@@ -0,0 +1,818 @@
+            last_raw_point = p1;
+
+            last_loop_check = p0; // check new segments for intersections
+
+            if(left >1) {  // осталась одна новая точка 
+                last_reduce = move_back(last_big_reduce, SIMPLIFY_MOVE_BACK); // move back by SIMPLIFY_MOVE_BACK points
+                last_big_reduce = p1;
+                DBG_PRINTVAR( left); DBG_PRINTVARLN( last_reduce);
+
+                in_loop_reduce=true;   // check newly formed segments
+            } else { // no new points
+                last_reduce = p1; // just from last point
+            }
+
+
+    DBG_PRINTVARLN( left);
+
+    DBG_PRINTVAR( removed);   DBG_PRINTVAR( last_reduce);  DBG_PRINTVARLN( last_big_reduce);
+        
+            if((num_points-removed)>=max_num_points) goto again; // we still don't have a room for new point
+        }
+    }
+    
+    // if enabled or we should to clear room for new points: we check one leg per point so number of free points should be more than not checked legs
+    //                  process active    enabled                number of not checked legs          number of free points       corrected by one packet
+    if(! was_reduce ) {
+    
+        // to exclude quadratic complexity we check intersection only one step in time
+
+        if(in_loop_reduce){      //  уже обрабатываем или отстали на нужное количество точек 
+
+            // check one leg from last_loop_check
+            p0 = last_loop_check;
+            p1 = move_forw(p0,1); // points can be not removed so we use relative coordinates
+
+         // check segment for track intersection
+            uint16_t sb=0;
+            
+            if(p1>p0){
+    DBG_PRINT("check loop ");     DBG_PRINTVAR(p0);    DBG_PRINTVARLN(p1);
+
+                sb = try_remove_loop(p0, p1);
+            }
+
+
+            if(sb) { // loop found
+                p1=sb-1;    // prev point as last checked
+
+                last_reduce=p1; // try to reduce from here
+                new_points +=  MIN_SIMPLIFY_POINTS; // force reduce on next point
+            
+                if(p1<2) p1=2; // but not earlier than 2
+            
+            
+                was_reduce = true; // points was removed
+            } else {
+                if(p1 >= hi_loop_border)  in_loop_reduce=false;
+            }
+ 
+            
+            last_loop_check = p1;                    
+        }
+    }
+    // do it only if was any reductions
+    if(was_reduce) {
+        squizze();      // remove bad points
+    }
+
+}
+
+
+
+
+/*
+   Determine the intersection point of two line segments (Paul Bourke)
+   Return FALSE if the lines don't intersect
+*/
+bool AP_WayBack::linesIntersect(const Point &p1, const Point &p2,
+                                const Point &p3, const Point &p4,
+                                Point *cross) // coordinates of intersection
+{
+   float denom,numera,numerb;
+
+   denom  = (p4.y-p3.y) * (p2.x-p1.x) - (p4.x-p3.x) * (p2.y-p1.y);
```

Comment:
Moreover that Cortex has "composite" commands like vfma which used when compiling such code and not used in case load-mul-store. This code turns to 
80a5a7a:       ee07 1a10       vmov    s14, r1
 80a5a7e:       6859            ldr     r1, [r3, #4]
 80a5a80:       681b            ldr     r3, [r3, #0]
 80a5a82:       ee05 3a90       vmov    s11, r3
 80a5a86:       6843            ldr     r3, [r0, #4]
 80a5a88:       ee08 1a10       vmov    s16, r1
 80a5a8c:       ee05 3a10       vmov    s10, r3
 80a5a90:       6801            ldr     r1, [r0, #0]
 80a5a92:       686b            ldr     r3, [r5, #4]
 80a5a94:       ee08 1a90       vmov    s17, r1
 80a5a98:       ee06 2a90       vmov    s13, r2
 80a5a9c:       ee07 3a90       vmov    s15, r3
 80a5aa0:       6829            ldr     r1, [r5, #0]
 80a5aa2:       ee75 5ae6       vsub.f32        s11, s11, s13
 80a5aa6:       ee35 9a67       vsub.f32        s18, s10, s15
 80a5aaa:       ee06 1a10       vmov    s12, r1
 80a5aae:       ee38 8a47       vsub.f32        s16, s16, s14
 80a5ab2:       ee36 6a68       vsub.f32        s12, s12, s17
 80a5ab6:       ee29 9a25       vmul.f32        s18, s18, s11
 80a5aba:       ee76 8ae8       vsub.f32        s17, s13, s17
 80a5abe:       ee35 7a47       vsub.f32        s14, s10, s14
 80a5ac2:       eea8 9a06       vfma.f32        s18, s16, s12
 80a5ac6:       4604            mov     r4, r0
 80a5ac8:       ee28 8a28       vmul.f32        s16, s16, s17
 80a5acc:       ee77 7ac5       vsub.f32        s15, s15, s10
 80a5ad0:       eea5 8a87       vfma.f32        s16, s11, s14
 80a5ad4:       ee67 8aa8       vmul.f32        s17, s15, s17
 80a5ad0:       eea5 8a87       vfma.f32        s16, s11, s14
 80a5ad4:       ee67 8aa8       vmul.f32        s17, s15, s17
 80a5ad8:       eeb0 0a48       vmov.f32        s0, s16
 80a5adc:       eee6 8a07       vfma.f32        s17, s12, s14

one load per element - and all calculations in registers, almost without any VMOV

But if classes used and results saved to class members then optimization limited to one line

pos_vector = Vector3f(packet.x * 100.0f, packet.y * 100.0f, -packet.z * 100.0f);

 8027df2:       ed9f 6a18       vldr    s12, [pc, #96]  ; 8027e54 
 8027df6:       ed9d 7a16       vldr    s14, [sp, #88]  ; 0x58
 8027dfa:       eddd 7a17       vldr    s15, [sp, #92]  ; 0x5c
 8027dfe:       eddd 6a15       vldr    s13, [sp, #84]  ; 0x54
 8027e02:       ee27 7a06       vmul.f32        s14, s14, s12
 8027e06:       ee67 7ac6       vnmul.f32       s15, s15, s12
 8027e0a:       ee66 6a86       vmul.f32        s13, s13, s12
 8027e12:       edcd 6a08       vstr    s13, [sp, #32]
 8027e16:       ed8d 7a09       vstr    s14, [sp, #36]  ; 0x24
 8027e1a:       edcd 7a0a       vstr    s15, [sp, #40]

but this line looks much better, sure 

Also GCC don't likes to remove class methods even they are "inline" so all cool .is_zero() turns to real calls with creation of temporary variable in stack:
if (delta_pos_xyz.is_zero()) {
 80a231a:       4628            mov     r0, r5
 80a231c:       edcd 0a06       vstr    s1, [sp, #24]
 80a2320:       edcd 7a07       vstr    s15, [sp, #28]
 80a2324:       f7b1 f912       bl      805354c <_ZNK7Vector3IfE7is_zeroEv>
 80a2328:       b938            cbnz    r0, 80a233a

and even const function much worse than #defines - we got real function instead of true constant:
float AP_Proximity_LightWareSF40C::distance_min() const
{
    return 0.20f;
}
 80a3290:       ed9f 0a01       vldr    s0, [pc, #4]    ; 80a3298 
 80a3294:       4770            bx      lr